src/EventManager.php

```diff
@@ -262,58 +264,72 @@ protected function triggerListeners(EventInterface $event, callable $callback =
             throw new Exception\RuntimeException('Event is missing a name; cannot trigger!');
         }
 
-        // Initial value of stop propagation flag should be false
-        $event->stopPropagation(false);
-
-        $responses = new ResponseCollection();
-
-        foreach ($this->getListenersByEventName($name) as $listener) {
-            $response = $listener($event);
-            $responses->push($response);
-
-            // If the event was asked to stop propagating, do so
-            if ($event->propagationIsStopped()) {
-                $responses->setStopped(true);
-                break;
+        // Merge all listeners by priority together with format
+        //   [
+        //     <int priority1> => [
+        //       [
+        //         [<callable listener1>, <callable listener2>, ...]
+        //       ],
+        //       ...
+        //     ],
+        //     <int priority2> => [
+        //       [
+        //         [<callable listener1>, <callable listener2>, ...]
+        //       ],
+        //       ...
+        //     ],
+        //     ...
+        //   ]
+        //
+        // PERFORMANCE NOTE:
+        //   This way pushing the list of listeners into a new generated list
+        //   helps us to not call "array_merge_[recursive]"
+        //   and reusing the list of listeners we already have
+        //   instead of iterating over it and generating a new one.
+        //   -> In result it improves performance by up to 25% even if it looks a bit strange.
+        $listOfListenersByPriority = isset($this->events[$name]) ? $this->events[$name] : [];
+        if (isset($this->events['*'])) {
+            foreach ($this->events['*'] as $p => $l) {
```

Comment:
I totally agreed - it's just hard to name it in a consistent way as it's basically a list of lists of lists of listeners and at one level the key is the name and at another level the priority. I'll try my best ;)